File: android/app/src/test/java/com/h19h29/naymnaymlevelup/rebuild/RebuildFeatureGateTest.java
package com.h19h29.naymnaymlevelup.rebuild;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import com.h19h29.naymnaymlevelup.BuildConfig;
import com.h19h29.naymnaymlevelup.rebuild.foundation.RebuildFeatureGate;
import java.util.List;
import org.junit.Test;

public final class RebuildFeatureGateTest {
    @Test
    public void defaultsToBuildConfigValue() {
        assertFalse(
                RebuildFeatureGate.INSTANCE.isEnabled(
                        BuildConfig.NATIVE_REBUILD_ENABLED,
                        null));
    }

    @Test
    public void explicitTrueOverrideWins() {
        assertTrue(RebuildFeatureGate.INSTANCE.isEnabled(false, "true"));
    }

    @Test
    public void caseInsensitiveTrueOverrideWins() {
        assertTrue(RebuildFeatureGate.INSTANCE.isEnabled(false, "TrUe"));
    }

    @Test
    public void nonTrueOverrideDoesNotEnableTheRebuild() {
        for (String overrideValue :
                List.of("false", "1", "yes", " true ", "", "TRUE!", "enabled")) {
            assertFalse(
                    overrideValue + " must not enable the rebuild",
                    RebuildFeatureGate.INSTANCE.isEnabled(false, overrideValue));
        }
    }

    @Test
    public void falseOverrideDoesNotDisableAnEnabledBuild() {
        assertTrue(RebuildFeatureGate.INSTANCE.isEnabled(true, "false"));
    }
}

File: android/app/src/test/java/com/h19h29/naymnaymlevelup/rebuild/RebuildFeatureGateTest.kt
package com.h19h29.naymnaymlevelup.rebuild

import com.h19h29.naymnaymlevelup.BuildConfig
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class RebuildFeatureGateTest {
    private val gateClass = Class.forName("com.h19h29.naymnaymlevelup.rebuild.foundation.RebuildFeatureGate")
    private val gate = gateClass.getField("INSTANCE").get(null)
    private val isEnabled = gateClass.getMethod(
        "isEnabled",
        Boolean::class.javaPrimitiveType,
        String::class.java,
    )

    @Test
    fun defaultsToBuildConfigValue() {
        assertFalse(isEnabled(BuildConfig.NATIVE_REBUILD_ENABLED, null))
    }

    @Test
    fun explicitTrueOverrideWins() {
        assertTrue(isEnabled(false, "true"))
    }

    @Test
    fun caseInsensitiveTrueOverrideWins() {
        assertTrue(isEnabled(false, "TrUe"))
    }

    @Test
    fun nonTrueOverrideDoesNotEnableTheRebuild() {
        listOf("false", "1", "yes", " true ", "", "TRUE!", "enabled").forEach { overrideValue ->
            assertFalse("$overrideValue must not enable the rebuild", isEnabled(false, overrideValue))
        }
    }

    @Test
    fun falseOverrideDoesNotDisableAnEnabledBuild() {
        assertTrue(isEnabled(true, "false"))
    }

    private fun isEnabled(defaultValue: Boolean, overrideValue: String?): Boolean =
        isEnabled.invoke(gate, defaultValue, overrideValue) as Boolean
}
